pkg/ctl/create/cluster.go

```diff
@@ -273,6 +274,18 @@ func doCreateCluster(cmd *cmdutils.Cmd, ngFilter *filter.NodeGroupFilter, params
 		postClusterCreationTasks.Append(preNodegroupAddons)
 	}
 
+	// collect IRSA outputs from cfn
+	var irsas []outputs.IRSA
+	go func() {
```

Comment:
I'm fairly certain that this will not work or works coincidentally, but I'll have to confirm that. This looks like a race condition between when things finish / start up. But I'll examine that in detail later. Can't this be sequential to when he cloud formation finishes or a post finish task like other tasks that we have? For example karpenter is installs once output is available.

Reply:
This code works fine and there is no race condition. This goroutine has to exist at this point for the channel output to receive and collect the IRSA at the other end of the channel. These irsa are collected in a slice and so no issue there. There are existing goroutines that were running in parellel cfn stacks to create the IRSA -- I just added to those goroutines the logic to send their respective IRSA info to this channel for collection in the above new goroutine. It all works fine. Here is console log output and I broke out the collected irsa info --
```
Starting: /Users/gerjohns/go/bin/dlv dap --check-go-version=false --listen=127.0.0.1:65456 --log-dest=3 from /Users/gerjohns/projects/jerryjs/eksctl/cmd/eksctl
DAP server listening at: 127.0.0.1:65456
Type 'dlv help' for list of commands.
2022-04-08 13:53:05 [ℹ]  eksctl version 0.93.0-dev
2022-04-08 13:53:05 [ℹ]  using region us-east-2
2022-04-08 13:53:05 [ℹ]  subnets for us-east-2c - public:192.168.0.0/19 private:192.168.96.0/19
2022-04-08 13:53:05 [ℹ]  subnets for us-east-2b - public:192.168.32.0/19 private:192.168.128.0/19
2022-04-08 13:53:05 [ℹ]  subnets for us-east-2a - public:192.168.64.0/19 private:192.168.160.0/19
2022-04-08 13:53:05 [ℹ]  nodegroup "ng-1" will use "ami-0656dd273bd6e9a2f" [AmazonLinux2/1.21]
2022-04-08 13:53:07 [ℹ]  using Kubernetes version 1.21
2022-04-08 13:53:07 [ℹ]  creating EKS cluster "foo" in "us-east-2" region with managed nodes
2022-04-08 13:53:07 [ℹ]  1 nodegroup (ng-1) was included (based on the include/exclude rules)
2022-04-08 13:53:07 [ℹ]  will create a CloudFormation stack for cluster itself and 0 nodegroup stack(s)
2022-04-08 13:53:07 [ℹ]  will create a CloudFormation stack for cluster itself and 1 managed nodegroup stack(s)
2022-04-08 13:53:07 [ℹ]  if you encounter any issues, check CloudFormation console or try 'eksctl utils describe-stacks --region=us-east-2 --cluster=foo'
2022-04-08 13:53:07 [ℹ]  Kubernetes API endpoint access will use default of {publicAccess=true, privateAccess=false} for cluster "foo" in "us-east-2"
2022-04-08 13:53:07 [ℹ]  CloudWatch logging will not be enabled for cluster "foo" in "us-east-2"
2022-04-08 13:53:07 [ℹ]  you can enable it with 'eksctl utils update-cluster-logging --enable-types={SPECIFY-YOUR-LOG-TYPES-HERE (e.g. all)} --region=us-east-2 --cluster=foo'
2022-04-08 13:53:07 [ℹ]  
2 sequential tasks: { create cluster control plane "foo", 
    2 sequential sub-tasks: { 
        4 sequential sub-tasks: { 
            wait for control plane to become ready,
            associate IAM OIDC provider,
            4 parallel sub-tasks: { 
                2 sequential sub-tasks: { 
                    create IAM role for serviceaccount "amazon-cloudwatch/fluent-bit",
                    create serviceaccount "amazon-cloudwatch/fluent-bit",
                },
                2 sequential sub-tasks: { 
                    create IAM role for serviceaccount "amazon-cloudwatch/cloudwatch-agent",
                    create serviceaccount "amazon-cloudwatch/cloudwatch-agent",
                },
                2 sequential sub-tasks: { 
                    create IAM role for serviceaccount "kube-system/aws-load-balancer-controller",
                    create serviceaccount "kube-system/aws-load-balancer-controller",
                },
                2 sequential sub-tasks: { 
                    create IAM role for serviceaccount "kube-system/aws-node",
                    create serviceaccount "kube-system/aws-node",
                },
            },
            restart daemonset "kube-system/aws-node",
        },
        create managed nodegroup "ng-1",
    } 
}
2022-04-08 13:53:07 [ℹ]  building cluster stack "eksctl-foo-cluster"
2022-04-08 13:53:09 [ℹ]  deploying stack "eksctl-foo-cluster"
2022-04-08 13:53:39 [ℹ]  waiting for CloudFormation stack "eksctl-foo-cluster"
2022-04-08 13:54:09 [ℹ]  waiting for CloudFormation stack "eksctl-foo-cluster"
2022-04-08 13:55:10 [ℹ]  waiting for CloudFormation stack "eksctl-foo-cluster"
2022-04-08 13:56:11 [ℹ]  waiting for CloudFormation stack "eksctl-foo-cluster"
2022-04-08 13:57:12 [ℹ]  waiting for CloudFormation stack "eksctl-foo-cluster"
2022-04-08 13:58:13 [ℹ]  waiting for CloudFormation stack "eksctl-foo-cluster"
2022-04-08 13:59:13 [ℹ]  waiting for CloudFormation stack "eksctl-foo-cluster"
2022-04-08 14:00:15 [ℹ]  waiting for CloudFormation stack "eksctl-foo-cluster"
2022-04-08 14:01:15 [ℹ]  waiting for CloudFormation stack "eksctl-foo-cluster"
2022-04-08 14:02:17 [ℹ]  waiting for CloudFormation stack "eksctl-foo-cluster"
2022-04-08 14:03:17 [ℹ]  waiting for CloudFormation stack "eksctl-foo-cluster"
2022-04-08 14:04:19 [ℹ]  waiting for CloudFormation stack "eksctl-foo-cluster"
2022-04-08 14:05:19 [ℹ]  waiting for CloudFormation stack "eksctl-foo-cluster"
2022-04-08 14:07:29 [ℹ]  building iamserviceaccount stack "eksctl-foo-addon-iamserviceaccount-kube-system-aws-node"
2022-04-08 14:07:29 [ℹ]  building iamserviceaccount stack "eksctl-foo-addon-iamserviceaccount-kube-system-aws-load-balancer-controller"
2022-04-08 14:07:29 [ℹ]  building iamserviceaccount stack "eksctl-foo-addon-iamserviceaccount-amazon-cloudwatch-cloudwatch-agent"
2022-04-08 14:07:29 [ℹ]  building iamserviceaccount stack "eksctl-foo-addon-iamserviceaccount-amazon-cloudwatch-fluent-bit"
2022-04-08 14:07:31 [ℹ]  deploying stack "eksctl-foo-addon-iamserviceaccount-amazon-cloudwatch-cloudwatch-agent"
2022-04-08 14:07:31 [ℹ]  waiting for CloudFormation stack "eksctl-foo-addon-iamserviceaccount-amazon-cloudwatch-cloudwatch-agent"
2022-04-08 14:07:31 [ℹ]  deploying stack "eksctl-foo-addon-iamserviceaccount-kube-system-aws-node"
2022-04-08 14:07:31 [ℹ]  waiting for CloudFormation stack "eksctl-foo-addon-iamserviceaccount-kube-system-aws-node"
2022-04-08 14:07:31 [ℹ]  deploying stack "eksctl-foo-addon-iamserviceaccount-amazon-cloudwatch-fluent-bit"
2022-04-08 14:07:31 [ℹ]  waiting for CloudFormation stack "eksctl-foo-addon-iamserviceaccount-amazon-cloudwatch-fluent-bit"
2022-04-08 14:07:31 [ℹ]  deploying stack "eksctl-foo-addon-iamserviceaccount-kube-system-aws-load-balancer-controller"
2022-04-08 14:07:31 [ℹ]  waiting for CloudFormation stack "eksctl-foo-addon-iamserviceaccount-kube-system-aws-load-balancer-controller"
2022-04-08 14:07:47 [ℹ]  waiting for CloudFormation stack "eksctl-foo-addon-iamserviceaccount-kube-system-aws-load-balancer-controller"
2022-04-08 14:07:48 [ℹ]  waiting for CloudFormation stack "eksctl-foo-addon-iamserviceaccount-amazon-cloudwatch-cloudwatch-agent"
2022-04-08 14:07:50 [ℹ]  waiting for CloudFormation stack "eksctl-foo-addon-iamserviceaccount-kube-system-aws-node"
2022-04-08 14:07:50 [ℹ]  waiting for CloudFormation stack "eksctl-foo-addon-iamserviceaccount-amazon-cloudwatch-fluent-bit"
2022-04-08 14:07:51 [ℹ]  created namespace "amazon-cloudwatch"
2022-04-08 14:07:51 [ℹ]  created serviceaccount "amazon-cloudwatch/fluent-bit"
2022-04-08 14:08:03 [ℹ]  waiting for CloudFormation stack "eksctl-foo-addon-iamserviceaccount-kube-system-aws-load-balancer-controller"
2022-04-08 14:08:06 [ℹ]  waiting for CloudFormation stack "eksctl-foo-addon-iamserviceaccount-amazon-cloudwatch-cloudwatch-agent"
2022-04-08 14:08:06 [ℹ]  waiting for CloudFormation stack "eksctl-foo-addon-iamserviceaccount-kube-system-aws-node"
2022-04-08 14:08:06 [ℹ]  created serviceaccount "amazon-cloudwatch/cloudwatch-agent"
2022-04-08 14:08:07 [ℹ]  serviceaccount "kube-system/aws-node" already exists
2022-04-08 14:08:07 [ℹ]  updated serviceaccount "kube-system/aws-node"
2022-04-08 14:08:20 [ℹ]  waiting for CloudFormation stack "eksctl-foo-addon-iamserviceaccount-kube-system-aws-load-balancer-controller"
2022-04-08 14:08:21 [ℹ]  created serviceaccount "kube-system/aws-load-balancer-controller"
2022-04-08 14:08:21 [ℹ]  daemonset "kube-system/aws-node" restarted
2022-04-08 14:08:22 [ℹ]  building managed nodegroup stack "eksctl-foo-nodegroup-ng-1"
2022-04-08 14:08:22 [ℹ]  deploying stack "eksctl-foo-nodegroup-ng-1"
2022-04-08 14:08:22 [ℹ]  waiting for CloudFormation stack "eksctl-foo-nodegroup-ng-1"
2022-04-08 14:08:39 [ℹ]  waiting for CloudFormation stack "eksctl-foo-nodegroup-ng-1"
2022-04-08 14:08:59 [ℹ]  waiting for CloudFormation stack "eksctl-foo-nodegroup-ng-1"
2022-04-08 14:09:17 [ℹ]  waiting for CloudFormation stack "eksctl-foo-nodegroup-ng-1"
2022-04-08 14:09:34 [ℹ]  waiting for CloudFormation stack "eksctl-foo-nodegroup-ng-1"
2022-04-08 14:09:54 [ℹ]  waiting for CloudFormation stack "eksctl-foo-nodegroup-ng-1"
2022-04-08 14:10:14 [ℹ]  waiting for CloudFormation stack "eksctl-foo-nodegroup-ng-1"
2022-04-08 14:10:30 [ℹ]  waiting for CloudFormation stack "eksctl-foo-nodegroup-ng-1"
2022-04-08 14:10:46 [ℹ]  waiting for CloudFormation stack "eksctl-foo-nodegroup-ng-1"
2022-04-08 14:11:03 [ℹ]  waiting for CloudFormation stack "eksctl-foo-nodegroup-ng-1"
```
```
2022-04-08 14:11:03 [ℹ]  IRSAs collected: [{IAMRole:arn:aws:iam::741162512851:role/eksctl-foo-addon-iamserviceaccount-amazon-cl-Role1-1R9U1AZUVTSAI ServiceAccount:fluent-bit} {IAMRole:arn:aws:iam::741162512851:role/eksctl-foo-addon-iamserviceaccount-amazon-cl-Role1-1B8AQW6F6K83A ServiceAccount:cloudwatch-agent} {IAMRole:arn:aws:iam::741162512851:role/eksctl-foo-addon-iamserviceaccount-kube-syst-Role1-J4RM7O3H7KCV ServiceAccount:aws-node} {IAMRole:arn:aws:iam::741162512851:role/eksctl-foo-addon-iamserviceaccount-kube-syst-Role1-G8NZ5RV08SZC ServiceAccount:aws-load-balancer-controller}]
```
```
2022-04-08 14:11:03 [ℹ]  waiting for the control plane availability...
2022-04-08 14:11:03 [✔]  saved kubeconfig as "/Users/gerjohns/.kube/config"
2022-04-08 14:11:03 [ℹ]  no tasks
2022-04-08 14:11:03 [✔]  all EKS cluster resources for "foo" have been created
2022-04-08 14:11:05 [ℹ]  nodegroup "ng-1" has 2 node(s)
2022-04-08 14:11:05 [ℹ]  node "ip-192-168-113-175.us-east-2.compute.internal" is ready
2022-04-08 14:11:05 [ℹ]  node "ip-192-168-188-203.us-east-2.compute.internal" is ready
2022-04-08 14:11:05 [ℹ]  waiting for at least 2 node(s) to become ready in "ng-1"
2022-04-08 14:11:05 [ℹ]  nodegroup "ng-1" has 2 node(s)
2022-04-08 14:11:05 [ℹ]  node "ip-192-168-113-175.us-east-2.compute.internal" is ready
2022-04-08 14:11:05 [ℹ]  node "ip-192-168-188-203.us-east-2.compute.internal" is ready
2022-04-08 14:11:05 [ℹ]  gitops configuration detected, setting installer to Flux v2
2022-04-08 14:11:05 [ℹ]  ensuring v1 repo components not installed
2022-04-08 14:11:05 [ℹ]  running pre-flight checks
► checking prerequisites
✗ flux 0.27.4 <0.28.5 (new version is available, please upgrade)
✔ Kubernetes 1.21.9-eks-0d102a7 >=1.20.6-0
✔ prerequisites checks passed
2022-04-08 14:11:16 [ℹ]  bootstrapping Flux v2 into cluster
► cloning branch "master" from Git repository "ssh://APKA2ZEF2OHJ4LHU374T@git-codecommit.us-east-2.amazonaws.com/v1/repos/k8s-gitops"
✔ cloned repository
► generating component manifests
✔ generated component manifests
✔ component manifests are up to date
► installing components in "flux-system" namespace
✔ installed components
✔ reconciled components
► determining if source secret "flux-system/flux-system" exists
► generating source secret
✔ public key: ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAACAQCx7pN3y1M7YAGQKQEUOvvPigKOQTJfIsbOUtMUub/bwVHst3uPLiiKg7mN394wUpTaMXMsy7473luJA2VFDj7qP3Kz4TeL8/uGyXuRJvgd84msvst+70GhApPCq69iSQDAj1rRIu5ksDjOHL1o7H+38ohGd0Rj5j9v46sH+PdT4Xd+eWCAyLPpIUfByY/XmcIM8SXtD+to0eUJMru7tkgyLa/MkGp/i4+KwmKvyenAFqp6bQysV5sJlZysuvhlSR+VpbquY6n03XlYV6T9cmboKXS0EPePSJtq71NAra3NDtipdVKtoUwzxjV8v82BDB5p2M3bMVsruP3mJamWi0aHg1Uo+mL3yBNT1ZB4XJZ33iUSo5LIQ27IcE931dDTY7HMtOGC+7YFgf85QGDSzj+mA8ApxMTcp6MB7cjqLlRvW5lmHYdoMdPS04JjDAFyKFHs308AutCvx77aD7tzgyW3DgEngl2+bFxXi7O7nZggIxhFAtdxM+536XmLMw3U8I11ChLsuQD8SRJx/RpsBy8W88tTJJtcK3kCxbWnjFtluWvksKinMNloXN1u9uMpgzGFzODKrU5tTMQ/Fr/wHwqbrFKwkdA0X9Lz79HMBjOTLJBJJjSj0gFNBa3lNWUggc3RHYvwwiPB/oYKLKCFfZmwGhBP7EZUirtzGB4lVTv69w==
► applying source secret "flux-system/flux-system"
✔ reconciled source secret
► generating sync manifests
✔ generated sync manifests
✔ sync manifests are up to date
► applying sync manifests
✔ reconciled sync configuration
◎ waiting for Kustomization "flux-system/flux-system" to be reconciled
✔ Kustomization reconciled successfully
► confirming components are healthy
✔ helm-controller: deployment ready
✔ kustomize-controller: deployment ready
✔ notification-controller: deployment ready
✔ source-controller: deployment ready
✔ all components are healthy
2022-04-08 14:15:04 [✔]  Flux v2 installed successfully
2022-04-08 14:15:04 [✔]  see https://toolkit.fluxcd.io/ for usage instructions
Process 20178 has exited with status 0
Detaching
dlv dap (19712) exited with code: 0
```
It is this IRSA collection that is used to create the configmap and applied to the cluster. Again, this logic is only called for a new cluster. So this happens in the new installer.InjectIRSA() which has been added that generates and applies this configmap. Here is an example of that applied configmap --
```
k get cm irsa-from-create-cluster -n flux-system -o yaml
apiVersion: v1
data:
  aws_load_balancer_controller_role_arn: arn:aws:iam::741162512851:role/eksctl-foo-addon-iamserviceaccount-kube-syst-Role1-1GC2CELFMSWHK
  aws_node_role_arn: arn:aws:iam::741162512851:role/eksctl-foo-addon-iamserviceaccount-kube-syst-Role1-MZXIX63SFIPT
  cloudwatch_agent_role_arn: arn:aws:iam::741162512851:role/eksctl-foo-addon-iamserviceaccount-amazon-cl-Role1-1G5MXY2LM7FXX
  fluent_bit_role_arn: arn:aws:iam::741162512851:role/eksctl-foo-addon-iamserviceaccount-amazon-cl-Role1-1LPRXNE8VH03L
kind: ConfigMap
metadata:
  creationTimestamp: "2022-04-08T14:37:50Z"
  name: irsa-from-create-cluster
  namespace: flux-system
  resourceVersion: "2032"
  uid: ae267ad5-72a4-4bef-98f8-c4b1589b8b5e
```
Subsequently, the existing installer.Run() is calling bootstrap which will use this configmap info if someone chooses to in their kustomize/flux kustomize. For example, the following is how I use this configmap in the container-insights example --
```
apiVersion: kustomize.toolkit.fluxcd.io/v1beta2
kind: Kustomization
metadata:
  name: infrastructure-container-insights
  namespace: flux-system
spec:
  interval: 10m0s
  dependsOn:
    - name: flux-system
  sourceRef:
    kind: GitRepository
    name: flux-system
  path: ./infrastructure/container-insights
  prune: true
  validation: client
  patches:
  - patch: |-
      - op: add
        path: "/metadata/annotations"
        value:
          eks.amazonaws.com/role-arn: ${cloudwatch_agent_role_arn}
    target:
      kind: ServiceAccount
      name: cloudwatch-agent
      namespace: amazon-cloudwatch
  - patch: |-
      - op: add
        path: "/metadata/annotations"
        value:
          eks.amazonaws.com/role-arn: ${fluent_bit_role_arn}
    target:
      kind: ServiceAccount
      name: fluent-bit
      namespace: amazon-cloudwatch
  postBuild:
    substitute:
      ClusterName: "test"
      RegionName: "us-east-2"
      FluentBitHttpPort: '"2020"'
      FluentBitHttpServer: '"On"'
      FluentBitReadFromHead: '"Off"'
      FluentBitReadFromTail: '"On"'
    substituteFrom:
      - kind: ConfigMap
        name: irsa-from-create-cluster
```
You can see the **substituteFrom irsa-from-create-cluster configmap** and the **fluent_bit_role_arn** and **cloudwatch_agent_role_arn** variables that are substituted.

And here is the final cluster showing pods of interest for cloudwatch-agent and fluent-bit that are associated with the above flux kustomization --
```
k get po -A
NAMESPACE           NAME                                            READY   STATUS    RESTARTS   AGE
amazon-cloudwatch   cloudwatch-agent-7bncz                          1/1     Running   0          37s
amazon-cloudwatch   cloudwatch-agent-x7fqt                          1/1     Running   0          37s
amazon-cloudwatch   fluent-bit-b2sbk                                1/1     Running   0          37s
amazon-cloudwatch   fluent-bit-gcv59                                1/1     Running   0          37s
cert-manager        cert-manager-7976c8df8-kk4hg                    1/1     Running   0          32s
cert-manager        cert-manager-cainjector-7bfd6d7dd9-kkjg8        1/1     Running   0          32s
cert-manager        cert-manager-webhook-6d66b4cdd-sf6nc            1/1     Running   0          32s
flux-system         helm-controller-5bfd65cd5f-lbjx2                1/1     Running   0          87s
flux-system         kustomize-controller-6f44c8d499-6xzml           1/1     Running   0          87s
flux-system         notification-controller-844df5f694-wctng        1/1     Running   0          87s
flux-system         source-controller-6b6c7bc4bb-slmgx              1/1     Running   0          87s
kube-system         aws-load-balancer-controller-8454f7fdd4-74psb   1/1     Running   0          5s
kube-system         aws-node-2sw2x                                  1/1     Running   0          4m41s
kube-system         aws-node-hqhpl                                  1/1     Running   0          4m38s
kube-system         coredns-f47955f89-2vs6c                         1/1     Running   0          14m
kube-system         coredns-f47955f89-mgk8z                         1/1     Running   0          14m
kube-system         kube-proxy-2r84s                                1/1     Running   0          4m38s
kube-system         kube-proxy-gl5c5                                1/1     Running   0          4m41s
```
I hope this helps make it more clear.

Reply:
PS -- thank you for looking at this @Skarlso

Reply:
I still don't understand why this needs to sit with the waiter. I don't think that's the right place for it. This is a post create task, isn't it? It needs output from the created stack, doesn't it? The stack has a post create hook. 

Things in these:

```go
	// obtain cluster credentials, write kubeconfig

	{ // post-creation action
```

You can call `DescribeStack` here on the created cluster stack, which returns the CF Stack and you can search for the IRSA in the output there. This doesn't have to be in a channel and in the Waiter where we don't collect and return anything. It just waits, so let's try and keep the functionality to `waiting` in there if that makes sense? Unless I completely misunderstand what you are trying to do. :D

Reply:
Yes, @Skarlso will look at that. Thanks